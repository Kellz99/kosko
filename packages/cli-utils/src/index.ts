export { type GlobalArguments, globalOptions, parse } from "./command";
export { CLIError, handleError } from "./error";
export { setupLogger } from "./logger";
