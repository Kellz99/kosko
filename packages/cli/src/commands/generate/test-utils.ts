import { Manifest } from "@kosko/generate";

export function createManifest(
  value: Partial<Omit<Manifest, "report">> = {}
): Manifest {
  return {
    position: { path: "", index: [] },
    issues: [],
    data: undefined,
    report() {},
    ...value
  };
}
