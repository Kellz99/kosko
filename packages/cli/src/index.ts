/**
 * @packageDocumentation
 * @module @kosko/cli
 */

import { parse } from "./cli/command";
import { rootCmd } from "./commands/root";

export { handleError } from "./cli/error";

/**
 * Runs CLI with the given arguments.
 */
export async function run(
  argv: string[] = process.argv.slice(2)
): Promise<void> {
  await parse(rootCmd, argv);
}
