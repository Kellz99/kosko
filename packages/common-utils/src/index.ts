/**
 * Common utilities used in Kosko.
 *
 * @packageDocumentation
 */

/**
 * Returns true when type of `value` is `object` and is not `null`, `undefined` or
 * an array.
 *
 * @public
 */
export function isRecord(
  value: unknown
): value is Record<string | symbol, unknown> {
  return value != null && typeof value === "object" && !Array.isArray(value);
}

/**
 * Transforms `input` into an array, or leave it as-is if `input` is already an array.
 *
 * @public
 */
export function toArray<T>(input: T | T[]): T[] {
  return Array.isArray(input) ? input : [input];
}

/**
 * Returns `code` of an error-like object.
 *
 * @public
 */
export function getErrorCode(err: unknown): string | undefined {
  if (isRecord(err) && typeof err.code === "string") {
    return err.code;
  }
}

/**
 * Returns the group part of an Kubernetes API version string.
 *
 * @public
 */
export function apiVersionToGroup(apiVersion: string): string {
  const index = apiVersion.lastIndexOf("/");

  return index === -1 ? "" : apiVersion.substring(0, index);
}

/**
 * @public
 */
export interface ManifestMeta {
  apiVersion: string;
  kind: string;
  name: string;
  namespace?: string;
}

/**
 * Returns metadata from a Kubernetes manifest. Returns `undefined` if `value`
 * is not a valid manifest.
 *
 * @public
 */
export function getManifestMeta(value: unknown): ManifestMeta | undefined {
  if (!isRecord(value)) return;

  const { apiVersion, kind, metadata } = value;

  if (
    typeof apiVersion !== "string" ||
    typeof kind !== "string" ||
    !isRecord(metadata) ||
    typeof metadata.name !== "string"
  ) {
    return;
  }

  return {
    apiVersion,
    kind,
    name: metadata.name,
    ...(typeof metadata.namespace === "string" && {
      namespace: metadata.namespace
    })
  };
}
