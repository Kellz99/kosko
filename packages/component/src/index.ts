export * from "./types";
export * from "./yaml";
