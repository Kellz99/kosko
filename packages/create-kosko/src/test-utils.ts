import { join, posix } from "node:path";
import glob from "fast-glob";
import { readFile } from "node:fs/promises";

export async function listAllFiles(
  dir: string
): Promise<Record<string, string>> {
  const paths = await glob("**/*", { cwd: dir });
  const files: Record<string, string> = {};

  for (const path of paths) {
    files[posix.normalize(path)] = await readFile(join(dir, path), "utf8");
  }

  return files;
}
