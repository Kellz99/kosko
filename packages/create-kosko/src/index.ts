import { createRootCommand, parse } from "@kosko/cli-utils";
import { argv } from "node:process";
import { version } from "../package.json";
import { command } from "./command";

export { handleError } from "@kosko/cli-utils";

function createCommand(args: readonly string[]) {
  return createRootCommand(args)
    .scriptName("create-kosko")
    .version(version)
    .command(command);
}

/**
 * Runs CLI with the given arguments.
 *
 * @public
 */
export async function run(
  args: readonly string[] = argv.slice(2)
): Promise<void> {
  await parse(createCommand(args), args);
}
