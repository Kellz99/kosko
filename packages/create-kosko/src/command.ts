import { mkdir, stat, writeFile } from "node:fs/promises";
import { dirname, join, relative, resolve, sep } from "node:path";
import pc from "picocolors";
import { type GlobalArguments, CLIError } from "@kosko/cli-utils";
import logger, { LogLevel } from "@kosko/log";
import isFolderEmpty from "./is-folder-empty";
import { File, Template } from "./templates/base";
import cjsTemplate from "./templates/cjs";
import tsTemplate from "./templates/ts";
import esmTemplate from "./templates/esm";
import tsEsmTemplate from "./templates/ts-esm";
import {
  getPackageManager,
  getInstallCommand,
  installDependencies
} from "./install";
import { getErrorCode } from "@kosko/common-utils";
import type { CommandModule } from "yargs";
import prompts from "prompts";
import isInteractive from "is-interactive";
import { exit, stdout } from "node:process";

function onPromptState({ aborted }: { aborted: boolean }) {
  if (aborted) {
    // Re-enable terminal cursor otherwise it will remain hidden after exit.
    stdout.write("\x1B[?25h");
    stdout.write("\n");
    exit(1);
  }
}

async function getPath(
  args: Pick<Arguments, "path" | "interactive">
): Promise<string | undefined> {
  if (typeof args.path === "string") return args.path;
  if (!args.interactive) return;

  const res = await prompts({
    onState: onPromptState,
    type: "text",
    name: "path",
    message: "Project path",
    initial: "kosko-project"
  });

  return res.path;
}

async function checkPath({
  path,
  force,
  interactive
}: {
  path: string;
  force?: boolean;
  interactive?: boolean;
}): Promise<void> {
  try {
    logger.log(LogLevel.Debug, `Checking stats of "${path}"`);
    const stats = await stat(path);

    if (!stats.isDirectory()) {
      throw new CLIError("Destination already exists and is not a directory", {
        output:
          "Destination already exists and is not a directory. Please delete it or try another path."
      });
    }

    if (force) {
      return;
    }

    if (await isFolderEmpty(path)) {
      logger.log(LogLevel.Trace, "Path can be initialized because it is empty");
      return;
    }

    const overwrite = await confirm({
      interactive,
      message: "Destination already exists. Do you want to overwrite it?",
      fallback: false
    });

    if (overwrite) return;

    throw new CLIError("Destination already exists", {
      output: `Destination already exists. Please empty the directory or rerun with "--force" to proceed.`
    });
  } catch (err) {
    if (getErrorCode(err) !== "ENOENT") throw err;

    logger.log(
      LogLevel.Trace,
      "Path can be initialized because it does not exist yet"
    );
  }
}

async function confirm({
  message,
  interactive,
  initial,
  fallback
}: {
  /**
   * Prompt message.
   */
  message: string;
  /**
   * Enable interactive mode. When the value is undefined, and interactive is
   * false, it will return the fallback value.
   */
  interactive?: boolean;
  /**
   * Default value for the prompt.
   */
  initial?: boolean;
  /**
   * Fallback value when interactive mode is disabled.
   */
  fallback: boolean;
}): Promise<boolean> {
  if (!interactive) return fallback;

  const res = await prompts({
    onState: onPromptState,
    type: "confirm",
    name: "confirm",
    message,
    initial
  });

  return res.confirm;
}

function resolvePath(cwd: string, path?: string): string {
  return path ? resolve(cwd, path) : cwd;
}

async function writeFiles(path: string, files: readonly File[]): Promise<void> {
  for (const file of files) {
    const filePath = join(path, file.path);

    logger.log(LogLevel.Debug, `Writing file: "${file.path}"`);
    await mkdir(dirname(filePath), { recursive: true });
    await writeFile(filePath, file.content);
  }
}

function getCDPath(cwd: string, path: string): string | undefined {
  if (cwd === path) return;

  if (path.startsWith(cwd + sep)) {
    return relative(cwd, path);
  }

  return path;
}

export interface Arguments extends GlobalArguments {
  force?: boolean;
  path?: string;
  typescript?: boolean;
  esm?: boolean;
  install?: boolean;
  packageManager?: string;
  interactive: boolean;
}

export const command: CommandModule<GlobalArguments, Arguments> = {
  command: "$0 [path]",
  describe: "Set up a new Kosko project",
  builder(argv) {
    // Don't set the default value if we are going to prompt the user. Set the
    // defaultDescription instead.
    return argv
      .positional("path", { type: "string", describe: "Path to initialize" })
      .option("force", {
        type: "boolean",
        describe: "Overwrite existing files",
        alias: "f"
      })
      .option("typescript", {
        type: "boolean",
        describe: "Generate TypeScript files",
        alias: "ts",
        defaultDescription: "true"
      })
      .option("esm", {
        type: "boolean",
        describe: "Generate ECMAScript module (ESM) files"
      })
      .option("install", {
        type: "boolean",
        describe: "Install dependencies automatically"
      })
      .option("package-manager", {
        type: "string",
        describe: "Package manager (npm, yarn, pnpm)",
        alias: "pm",
        defaultDescription: "auto"
      })
      .option("interactive", {
        type: "boolean",
        describe: "Run in interactive mode",
        default: isInteractive(),
        defaultDescription: "auto"
      });
  },
  async handler(args) {
    const path = resolvePath(args.cwd, await getPath(args));

    await checkPath({ path, force: args.force, interactive: args.interactive });

    args.typescript ??= await confirm({
      message: "Would you like to use TypeScript?",
      interactive: args.interactive,
      // Set to true because we prefer to use TypeScript.
      initial: true,
      fallback: true
    });

    args.esm ??= await confirm({
      interactive: args.interactive,
      message: "Would you like to use ECMAScript module (ESM)?",
      fallback: false
    });

    args.install ??= await confirm({
      interactive: args.interactive,
      message: "Would you like to install dependencies automatically?",
      // Set to true because we prefer to install dependencies
      // automatically.
      initial: true,
      // Set to false because users probably don't want to install dependencies
      // automatically when they are running in non-interactive mode.
      fallback: false
    });

    logger.log(LogLevel.Info, `Creating a Kosko project in "${path}"`);

    const template: Template = (() => {
      if (args.typescript) {
        return args.esm ? tsEsmTemplate : tsTemplate;
      }

      return args.esm ? esmTemplate : cjsTemplate;
    })();

    const packageManager = args.packageManager || getPackageManager();
    const runCmd = `${packageManager} run`;
    const { dependencies, devDependencies, files } = await template({ path });

    await writeFiles(path, files);

    const cdPath = getCDPath(args.cwd, path);
    let installSuccessful = false;

    if (args.install) {
      try {
        if (dependencies?.length) {
          await installDependencies({
            cwd: path,
            packageManager,
            dependencies
          });
        }

        if (devDependencies?.length) {
          await installDependencies({
            cwd: path,
            packageManager,
            dependencies: devDependencies,
            dev: true
          });
        }

        installSuccessful = true;
      } catch (err) {
        logger.log(LogLevel.Warn, "Install failed", { error: err });
      }
    }

    logger.log(
      LogLevel.Info,
      `Project created at "${path}"

Inside that directory, you can run several commands:
${[
  [
    `${runCmd} generate`,
    "Validate components and generate Kubernetes manifests."
  ],
  [`${runCmd} validate`, "Only validate components."]
]
  .map(([cmd, desc]) => `\n  ${pc.cyan(cmd)}\n    ${desc}`)
  .join("\n")}

We suggest that you begin by typing:

${[
  ...(cdPath ? [`cd ${cdPath}`] : []),
  ...(args.install && installSuccessful
    ? []
    : [
        dependencies?.length
          ? getInstallCommand({ packageManager, dependencies }).join(" ")
          : "",
        devDependencies?.length
          ? getInstallCommand({
              packageManager,
              dependencies: devDependencies,
              dev: true
            }).join(" ")
          : ""
      ]),
  `${runCmd} generate`
]
  .filter(Boolean)
  .map((line) => `  ${pc.cyan(line)}`)
  .join("\n")}`
    );
  }
};
