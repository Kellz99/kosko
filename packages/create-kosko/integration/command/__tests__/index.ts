import { TempDir, makeTempDir, projectRoot, runCLI } from "@kosko/test-utils";
import execa from "execa";
import { join } from "node:path";
import { listAllFiles } from "../../../src/test-utils";

const BIN_PATH = join(projectRoot, "packages/create-kosko/bin/create-kosko.js");

let tmpDir: TempDir;

function run(args: readonly string[], options?: execa.Options) {
  return runCLI(BIN_PATH, args, options);
}

beforeEach(async () => {
  tmpDir = await makeTempDir();
});

afterEach(() => tmpDir.cleanup());

describe("when target is an empty folder", () => {
  test("should write files to the target path", async () => {
    await run([tmpDir.path, "--install", "false"]);

    const files = await listAllFiles(tmpDir.path);
    expect(files).toMatchSnapshot();
  });
});
