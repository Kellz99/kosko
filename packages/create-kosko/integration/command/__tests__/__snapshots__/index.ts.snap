// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when target is an empty folder should write files to the target path 1`] = `
{
  "README.md": "This is a [Kosko] project bootstrapped with \`create-kosko\` command.

## Getting Started

Generate Kubernetes manifests with the following command.

\`\`\`sh
npm run generate
\`\`\`

The result will be printed in the console. You can pipe the output to \`kubectl\` to apply generated manifests on Kubernetes.

\`\`\`sh
npm run --silent generate | kubectl apply -f -
\`\`\`

When \`--env\` option is set, Kosko will read variables from environment files. Then you can fetch variables with \`@kosko/env\` package.

\`\`\`sh
npm run generate -- --env dev
\`\`\`

You can try editing files in \`components\` and \`environments\` folder and re-run \`npm run generate\` to see changes.

## Learn More

Check [Kosko docs](https://kosko.dev/docs/) for more information about Kosko.

[kosko]: https://kosko.dev/
",
  "components/nginx.js": "const { Deployment } = require("kubernetes-models/apps/v1/Deployment");
const { Service } = require("kubernetes-models/v1/Service");
const env = require("@kosko/env");

const params = env.component("nginx");

const name = "nginx";
const labels = { app: name };
const port = 80;

const deployment = new Deployment({
  metadata: {
    name,
    namespace: params.namespace
  },
  spec: {
    replicas: params.replicas,
    selector: {
      matchLabels: labels
    },
    template: {
      metadata: { labels },
      spec: {
        containers: [
          {
            name: "nginx",
            image: "nginx:stable",
            ports: [{ containerPort: port }],
            env: [{ name: "NGINX_PORT", value: \`\${port}\` }]
          }
        ]
      }
    }
  }
});

const service = new Service({
  metadata: {
    name,
    namespace: params.namespace
  },
  spec: {
    selector: labels,
    ports: [{ port }]
  }
});

module.exports = [deployment, service];
",
  "environments/dev/index.js": "module.exports = {
  namespace: "dev"
};
",
  "environments/dev/nginx.js": "module.exports = {
  replicas: 3
};
",
  "kosko.toml": "components = ["*"]
",
  "package.json": "{
  "scripts": {
    "generate": "kosko generate",
    "validate": "kosko validate"
  }
}
",
}
`;
